src/vmm/src/guest_config/aarch64/mod.rs

```diff
@@ -24,56 +25,62 @@ pub struct CpuConfiguration {
-            Some(ref cpu_template) => Self::with_template(vcpu, cpu_template),
-            None => Ok(Self::default()),
-        }
-    }
-
-    /// Creates new guest CPU config based on the provided template
-    fn with_template(vcpu: &VcpuFd, template: &CpuTemplateType) -> Result<Self, Error> {
-        match template {
-            CpuTemplateType::Custom(template) => Self::with_applied_template(vcpu, template),
-            CpuTemplateType::Static(StaticCpuTemplate::V1N1) => {
-                let template = static_cpu_templates::v1n1::v1n1();
-                Self::with_applied_template(vcpu, &template)
-            }
-            _ => unreachable!("Options other than V1N1 are invalid"),
-        }
-    }
-
-    /// Creates new guest CPU config based on the provided template
-    fn with_applied_template(vcpu: &VcpuFd, template: &CustomCpuTemplate) -> Result<Self, Error> {
-        let regs = template
-            .reg_modifiers
+    pub fn new(_vm: &Vm, vcpu: &KvmVcpu, template: &CustomCpuTemplate) -> Result<Self, Error> {
+        let reg_list = template.reg_list();
+        let regs = reg_list
             .iter()
-            .map(|modifier| {
-                vcpu.get_one_reg(modifier.addr)
-                    .map(|value| Aarch64Register {
-                        id: modifier.addr,
-                        value: modifier.bitmap.apply(value),
-                    })
-                    .map_err(|e| Error(ArchError::GetSysRegister(modifier.addr, e)))
+            .map(|id| {
+                vcpu.fd
+                    .get_one_reg(*id)
+                    .map(|value| Aarch64Register { id: *id, value })
+                    .map_err(|err| Error(ArchError::GetOneReg(*id, err)))
             })
             .collect::<Result<Vec<_>, Error>>()?;
 
         Ok(Self { regs })
     }
 
+    /// Creates new guest CPU config based on the provided template
+    pub fn apply_template(mut self, template: &CustomCpuTemplate) -> Result<Self, Error> {
+        for (modifier, reg) in template.reg_modifiers.iter().zip(self.regs.iter_mut()) {
+            reg.value = modifier.bitmap.apply(reg.value);
+        }
+        Ok(self)
+    }
+
     /// Returns ids of registers that are changed
     /// by this template
     pub fn register_ids(&self) -> Vec<u64> {
         self.regs.iter().map(|reg| reg.id).collect()
     }
+}
 
-    /// Applies cpu template to vcpu
-    /// Used inside Vcpu to configure it
-    pub fn apply(&self, vcpu: &VcpuFd) -> Result<(), ArchError> {
-        for Aarch64Register { id, value } in self.regs.iter() {
-            vcpu.set_one_reg(*id, *value)
-                .map_err(|error| ArchError::SetSysRegister(*id, error))?;
-        }
-        Ok(())
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::guest_config::templates::aarch64::{RegisterModifier, RegisterValueFilter};
+
+    #[test]
+    fn test_base_cpu_configuration() {
+        let kvm = crate::vstate::system::KvmContext::new().unwrap();
```

Comment:
This test does not test `CpuConfiguration::apply_template()`, instead test `CpuConfiguration::new()` (In this implementation, only `CpuConfiguration::new()` requires KVM access).

It is possible to move out KVM access completely from `CpuConfiguration::new()`, but it introduces arch-specific operations in `configure_system_for_boot()` function.

Reply:
Yep I understand. The point I wanted to make is that the `apply_template()` function is a good candidate for unit testing because of how discrete it is.

`new()` relying on KVM is completely reasonable to me. What I am referring to is that the unit test is working with KVM as well, which is not really what a unit test should be doing. This is more suited to an integration test.

Reply:
I see. Thanks for your explanation! Our current code base have many unit tests requiring KVM access, but should I put them into integration tests? I'd prefer to put tests near the actual code. But as Nikita commented, we might be able to put them into different places. What do you think?
https://github.com/firecracker-microvm/firecracker/pull/3610#discussion_r1168459752

Reply:
That is true that we already have a number of "unittests" relying on KVM. This is wrong, in my opinion. We should not call such tests unittests. I propose, in the short term, to stop adding such unittests and, in the long term, find a way to isolate the calls to KVM (possibly into traits) that will enable stubbing/mocking them in unittests. This will give us greater control on the environment we execute the unitests in, and also it will make our tests less fragile.
We may then still have some tests in Rust that call into KVM, but these tests will be integration tests.
This all is probably in line with earlier Matthew's point.

Reply:
The short version of my suggestion is to add a Rust integration test (integration tests are located as a peer to the src folder, so a new test file at `./src/vmm/tests/cpu_template_integration.rs` for example), and move the test function `test_base_cpu_configuration` to the new integration test file. That way we can still get code coverage reported on it. A new unit test can be added to `guest_config/aarch64/mod.rs` that confirms `apply_template()` returns sane results when given a valid template, but also validates that the function generates errors when given invalid templates.

--------------------------------

It is my personal preference that unassociated functions (not a part of the `CpuConfiguration` impl) would be better suited to dealing with KVM and vcpu references. something like a "ConfigurationManager" to take in host data and a CPU templates, and produce the guest's CPU configuration. That was my original layout but the code has been refactored considerably. I don't think this is in scope anymore though.

Reply:
Thanks for good suggestion. I'll try to put them into rust integration tests `./src/vmm/tests/cpu_template_integration.rs`!